File: right_view_bt.py
# O(N) TIME AND O(D) SPACE WHERE N IS NO.OF NODES AND D IS DIAMETER
from collections import deque
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        if root is None:
            return []
        output = []
        que = deque()
        que.append(root)
        while que:
            size = len(que)
            for i in range(size):
                popped = que.popleft()
                if i == size - 1:
                    output.append(popped.val)
                if popped.left is not None:
                    que.append(popped.left)
                if popped.right is not None:
                    que.append(popped.right)
        return output
        

File: right-view-bt.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

// TC: O(N)
// SC: O(N)

class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        if(root == null) return new ArrayList<>();
        
        Queue<TreeNode> queue = new LinkedList<>();
        List<Integer> al = new ArrayList<>();
        queue.add(root);
        
        int count = 0;
        while(!queue.isEmpty()) {
            count = 0;
            int size = queue.size();
            for(int i=0;i<size;i++) {
                TreeNode curr = queue.peek();
                count++;
                if(curr.left != null) queue.add(curr.left);
                if(curr.right != null) queue.add(curr.right);
                queue.poll();
                if(count == size) {
                    al.add(curr.val);
                }
            }
        }
        return al;
    }
}
